src/olympia/editors/management/commands/auto_approve.py

```diff
@@ -59,57 +66,77 @@ def handle(self, *args, **options):
+        self.successful_verdict = (
+            amo.WOULD_HAVE_BEEN_AUTO_APPROVED if self.dry_run
             else amo.AUTO_APPROVED)
 
+        self.stats = Counter()
+        qs = self.fetch_candidates()
+        self.stats['total'] = len(qs)
+
         for version in qs:
-            # Is the addon already locked by a reviewer ?
-            if get_reviewing_cache(version.addon.pk):
-                stats['locked'] += 1
-                continue
-
-            # If admin review or more information was requested, skip this
-            # version, let a human handle it.
-            if version.addon.admin_review or version.has_info_request:
-                stats['flagged'] += 1
-                continue
-
-            # Lock the addon for ourselves, no reviewer should touch it.
-            set_reviewing_cache(version.addon.pk, settings.TASK_USER_ID)
-
-            try:
-                log.info('Processing %s version %s...',
-                         unicode(version.addon.name), unicode(version.version))
-                summary, info = AutoApprovalSummary.create_summary_for_version(
-                    version, max_average_daily_users=max_average_daily_users,
-                    min_approved_updates=min_approved_updates,
-                    dry_run=dry_run)
-                log.info('Auto Approval for %s version %s: %s',
-                         unicode(version.addon.name),
-                         unicode(version.version),
-                         summary.get_verdict_display())
-                stats.update({k: int(v) for k, v in info.items()})
-                if summary.verdict == successful_verdict:
-                    stats['auto_approved'] += 1
-                # FIXME: implement auto-approve if verdict is amo.AUTO_APPROVED
-
-            except (AutoApprovalNotEnoughFilesError,
-                    AutoApprovalNoValidationResultError):
-                log.info(
-                    'Version %s was skipped either because it had no '
-                    'file or because it had no validation attached.', version)
-                stats['error'] += 1
-            finally:
-                clear_reviewing_cache(version.addon.pk)
-
-        self.log_final_summary(stats, dry_run=dry_run)
-
-    def log_final_summary(self, stats, dry_run=False):
+            self.process(version)
+
+        self.log_final_summary(self.stats)
+
+    @transaction.atomic
+    def process(self, version):
+        """Process a single version, figuring out if it should be auto-approved
+        and calling the approval code if necessary."""
+
+        # Is the addon already locked by a reviewer ?
+        if get_reviewing_cache(version.addon.pk):
+            self.stats['locked'] += 1
+            return
+
+        # If admin review or more information was requested, skip this
+        # version, let a human handle it.
+        if version.addon.admin_review or version.has_info_request:
+            self.stats['flagged'] += 1
+            return
```

Comment:
I wonder if we really should skip here and avoid gathering more stats about the approval process?